dpnp/backend/extensions/sycl_ext/histogram.cpp

```diff
@@ -0,0 +1,266 @@
+//   this list of conditions and the following disclaimer.
+// - Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+#include <algorithm>
+#include <complex>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+// dpctl tensor headers
+#include "dpctl4pybind11.hpp"
+#include "utils/type_dispatch.hpp"
+
+#include "histogram.hpp"
+#include "histogram_common.hpp"
+
+namespace dpctl_td_ns = dpctl::tensor::type_dispatch;
+using dpctl::tensor::usm_ndarray;
+
+using namespace histogram;
+
+namespace
+{
+
+template <typename T, typename BinsT, typename HistType = size_t>
+static sycl::event histogram_impl(sycl::queue exec_q,
+                                  const void *vin,
+                                  const void *vbins_edges,
+                                  const void *vweights,
+                                  void *vout,
+                                  const size_t bins_count,
+                                  const size_t size,
+                                  const std::vector<sycl::event> &depends)
+{
+    const T *in = static_cast<const T *>(vin);
+    const BinsT *bins_edges = static_cast<const BinsT *>(vbins_edges);
+    const HistType *weights = static_cast<const HistType *>(vweights);
+    HistType *out = static_cast<HistType *>(vout);
+
+    auto device = exec_q.get_device();
+
+    uint32_t local_size =
+        device.is_cpu()
+            ? 256
+            : device.get_info<sycl::info::device::max_work_group_size>();
+
+    uint32_t WorkPI = 128; // empirically found number
+    auto global_size = Align(CeilDiv(size, WorkPI), local_size);
+
+    auto nd_range =
+        sycl::nd_range(sycl::range<1>(global_size), sycl::range<1>(local_size));
+
+    return exec_q.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(depends);
+        uint32_t dims = 1;
```

Comment:
Is there a use case when `dims` might differ from 1?

Reply:
for histogram - no. For histogramdd - yes. But they uses the same common kernel